lib/posix-process/signal/signal.h

```diff
@@ -0,0 +1,243 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#ifndef __UK_SIGNAL_H__
+#define __UK_SIGNAL_H__
+
+#include <limits.h>
+#include <signal.h>
+
+#include <uk/alloc.h>
+#include <uk/list.h>
+#include <uk/process.h>
+#include <uk/refcount.h>
+#include <uk/semaphore.h>
+
+#include "process.h"
+#include "sigset.h"
+
+#define SIG_ARRAY_COUNT		_NSIG /* SIGRTMAX + 1 */
+
+/* Check if signal number is valid */
+#define IS_VALID(_signum)						\
+	((_signum) > 0 && (_signum) <= SIGRTMAX)
+
+/* Check whether signal number is pending in sigqueue */
+#define IS_PENDING(_sigqueue, _signum)					\
+	uk_sigismember(&(_sigqueue).pending, (_signum))
+
+/* Set signal number in sigqueue */
+#define SET_PENDING(_sigqueue, _signum)					\
+	uk_sigaddset(&(_sigqueue).pending, (_signum))
+
+/* Clear a signal from a sigqueue */
+#define RESET_PENDING(_sigqueue, _signum)				\
+	uk_sigdelset(&(_sigqueue).pending, (_signum))
+
+/* Check if posix thread masks signal number */
+#define IS_MASKED(_t, _signum)						\
+	uk_sigismember(&(_t)->signal->mask, (_signum))
+
+/* Sets a posix thread's mask of the given signal number */
+#define SET_MASKED(_t, _signum)						\
+	uk_sigaddset(&(_t)->signal->mask, (_signum))
+
+/* Check if signal is ignored by a process. A signal is ignored if it
+ * is explicitly set to IGNORE by sigaction, or if it is set to its
+ * default action and that action is IGNORE.
+ */
+#define IS_IGNORED(_pproc, _signum)					  \
+	((KERN_SIGACTION(_pproc, _signum)->ks_handler == SIG_IGN) ||\
+	 (KERN_SIGACTION(_pproc, _signum)->ks_handler == SIG_DFL && \
+	  UK_BIT(_signum) & SIGACT_IGN_MASK))
+
+/* Allocate and return `struct uk_signal` or NULL */
+#define uk_signal_alloc(_alloc)						\
+	uk_zalloc(_alloc, sizeof(struct uk_signal))
+
+/* Free a previously allocated `struct uk_signal` */
+#define uk_signal_free(_alloc, _sig)					\
+	uk_free(_alloc, _sig)
+
+/* Helper to get pointer to kern_sigaction from process,
+ * for a given signal
+ */
+#define KERN_SIGACTION(_proc, _signum)	({				\
```

Comment:
I don't see any other subscope macro in the codebase with the `({` on the same line as the `#define`. Should we move this down one line for consistency instead?

Reply:
Okay, as I'm updating this file for https://github.com/unikraft/unikraft/pull/1248#discussion_r1961269344 anyway. But otherwise I wouldn't find this worth a review comment.